### DISCUSSION TITLE
integrations/onnx/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8340] [Docs]
# integrations/onnx/

Explore how to improve your Ultralytics YOLOv8 model's performance and interoperability using the ONNX (Open Neural Network Exchange) export format that is suitable for diverse hardware and software environments.

https://docs.ultralytics.com/integrations/onnx/

**Comment by frtendero** [comment]
I'm integrating pre and post processing operations into the exported ONNX model using the [onnx extensions package](https://onnxruntime.ai/docs/extensions/). I could do it for object detection following an [example](https://github.com/microsoft/onnxruntime-extensions/blob/main/onnxruntime_extensions/tools/add_pre_post_processing_to_model.py#L90) listed in their repo. Is there any reference on which operations are required for Segmentation and Classification models? (pre and post processing). Thanks!

**Reply by glenn-jocher** [reply]
Hey there! 👋

For integrating pre and post-processing operations into ONNX models for segmentation and classification tasks, you can follow a similar approach to what you've done with object detection. Generally, pre-processing involves resizing the input image to the model's expected dimensions and normalizing pixel values. Post-processing for segmentation might include applying a threshold to the output mask and extracting contours, while for classification, it typically involves selecting the class with the highest probability.

Here's a quick example for a classification model:

```python
# Pre-processing: Resize and normalize
input_image = cv2.resize(input_image, (model_input_width, model_input_height))
input_image = input_image / 255.0  # Assuming normalization to [0, 1]

# Post-processing: Argmax for classification
output_probabilities = model.run(input_image)
predicted_class = np.argmax(output_probabilities)
```

For segmentation, the post-processing might look like this:

```python
# Post-processing: Threshold and extract mask
output_mask = model.run(input_image)
_, thresholded_mask = cv2.threshold(output_mask, threshold_value, max_value, cv2.THRESH_BINARY)
contours, _ = cv2.findContours(thresholded_mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
```

Remember to adjust the `threshold_value` and `max_value` as needed for your specific model and application.

Hope this helps! If you need more detailed guidance, feel free to check out our [ONNX Export for YOLOv8 Models](https://docs.ultralytics.com/integrations/onnx/) documentation. 😊

Happy coding!

**Reply by frtendero** [reply]
Thanks for your answer Glenn. The difficulties I had were more about using onnx extensions to insert the processing in the onnx segmentation model file itself. Specially for the post-processing of the output (batch_size, 37, 8400): doing the required split and applying non max supression as well as the according coordinates transformation. **[This example from ONNX Extensions tutorials for yolov8 pose detection](https://github.com/microsoft/onnxruntime-extensions/blob/main/tutorials/yolov8_pose_e2e.py)** helped a lot as a basis for my segmentation problem.

**Reply by glenn-jocher** [reply]
Hey there! 😊

It's great to hear that the ONNX Extensions example for YOLOv8 pose detection was helpful for your segmentation task. For post-processing the segmentation output `(batch_size, 37, 8400)`, you're on the right track with splitting and applying non-max suppression (NMS), as well as transforming the coordinates.

While we don't have a specific example for segmentation in the ONNX Extensions repo, the general idea would be to:

1. Reshape or split your output tensor to separate the class predictions from the segmentation masks.
2. Apply a threshold to the segmentation masks to create binary masks.
3. Use NMS to filter overlapping masks, if your task involves instances of objects.
4. Convert the mask pixel indices back to your original image coordinates.

For classification models, the post-processing typically involves:
- Applying a softmax function to the output logits to get probabilities.
- Using `argmax` to find the predicted class.

Here's a pseudo-code snippet for segmentation post-processing:

```python
# Assuming 'output' is your model's raw output tensor
class_preds, masks = output[..., :num_classes], output[..., num_classes:]
masks = masks > segmentation_threshold  # Apply threshold
selected_indices = nms(masks, scores=class_preds, iou_threshold=iou_thresh)  # NMS
final_masks = masks[selected_indices]
```

And for classification post-processing:

```python
# Assuming 'output' is your model's raw output logits
probabilities = softmax(output)
predicted_class = argmax(probabilities)
```

Remember to tailor these snippets to fit your model's specific output format and the ONNX Extensions API. If you need more guidance, our [ONNX Export for YOLOv8 Models](https://docs.ultralytics.com/integrations/onnx/) documentation might provide additional insights.

Keep up the great work, and if you have any more questions, feel free to reach out! 🚀

**Comment by assia855** [comment]
Hi @frtendero  & @glenn-jocher  I'm trying to do the same thing. I would like to integrate the pre and post processing in my custom segmentation yolo model, but I don't understand how can I achieve that. Can you please, explain to me how can I  achieve that?

**Reply by glenn-jocher** [reply]
Hey @assia855! 😊 Integrating pre and post-processing directly into your custom YOLO segmentation model for ONNX export can streamline your deployment process. Here's a quick guide on how to achieve that:

1. **Pre-processing**: Ensure your model's input layer is designed to accept raw images in the expected format (e.g., RGB, BGR) and dimensions. You might need to adjust your model's first layer to perform resizing, normalization, etc., directly within the model.

2. **Post-processing**: For segmentation models, post-processing often involves converting the model's output tensors into meaningful segmentation maps. You can add additional layers at the end of your model that perform tasks like thresholding, non-maximum suppression (NMS), or generating segmentation masks from logits.

Here's a simple example of adding a sigmoid layer for binary segmentation tasks in PyTorch, which you can adapt for your needs:

```python
import torch
import torch.nn as nn

class CustomSegmentationModel(nn.Module):
    def __init__(self, base_model):
        super(CustomSegmentationModel, self).__init__()
        self.base_model = base_model  # Your pre-trained segmentation model
        self.sigmoid = nn.Sigmoid()  # Sigmoid layer for binary segmentation

    def forward(self, x):
        x = self.base_model(x)  # Pass input through the base model
        x = self.sigmoid(x)  # Apply sigmoid activation to get segmentation mask
        return x
```

After integrating these layers, you can export your model to ONNX as usual:

```python
model = CustomSegmentationModel(your_base_model)
model.eval()  # Set the model to evaluation mode
dummy_input = torch.randn(1, 3, 640, 640)  # Example input tensor
torch.onnx.export(model, dummy_input, "custom_segmentation_model.onnx", opset_version=11)
```

This approach allows you to encapsulate all necessary pre and post-processing steps within the model, making deployment more straightforward. Remember to adjust the input dimensions and processing steps according to your specific requirements.

Hope this helps! If you have further questions, feel free to ask. 🚀

**Reply by assia855** [reply]
Thank you so much @glenn-jocher  for taking time to explain me the process. I will try to follow your suggestions. Thanks again

**Reply by glenn-jocher** [reply]
Hey @assia855! 🌟

I'm thrilled to hear you're diving into integrating pre and post-processing with your custom segmentation YOLO model! Here's a quick rundown to get you started:

- **Pre-processing**: Embed this directly into your model's first layer if possible. This could include resizing, normalization, etc. For example, if you're using PyTorch:

```python
class CustomModel(nn.Module):
    def __init__(self):
        super(CustomModel, self).__init__()
        # Your model layers here
        self.conv1 = nn.Conv2d(3, 16, 3, stride=2, padding=1)
        # Add more layers

    def forward(self, x):
        # Pre-processing: Normalize x
        x = (x - 0.5) / 0.5
        # Forward pass through layers
        x = self.conv1(x)
        # More layers
        return x
```

- **Post-processing**: This might involve converting logits to segmentation masks or applying thresholding. You can add layers at the end of your model for this or handle it outside the model depending on your deployment needs.

Exporting your model with these integrated steps to ONNX is straightforward:

```python
import torch.onnx

# Assuming 'model' is your custom model instance
dummy_input = torch.randn(1, 3, 640, 480)  # Example input tensor
torch.onnx.export(model, dummy_input, "model_with_pre_post.onnx", opset_version=12)
```

This approach ensures your model is ready for deployment with minimal external processing required, making it more efficient and easier to manage. 🚀

If you have any more questions or need further assistance, feel free to reach out. Happy coding!

**Reply by assia855** [reply]
Thanks so much for your response @glenn-jocher  I was thinking to convert the functions: preprocessing and post processing  in [YOLOv8-Segmentation-ONNXRuntime-Python](https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py) to pytorch and merge them one by one with the model. I'm struggling with the conversion but I think it is the right way since the script is working really good with my custom model.

**Reply by glenn-jocher** [reply]
Hey @assia855 & @assia855! 🌟

It's fantastic to hear about your progress and plans to integrate pre and post-processing directly into your custom YOLO segmentation model for ONNX export. Converting these functions to PyTorch and merging them with your model is indeed a solid approach, especially since the script is performing well with your custom model.

For the pre-processing conversion, you can include normalization and resizing directly in your model's `forward()` method. Here's a quick example for normalization:

```python
def forward(self, x):
    # Normalize x
    x = (x - 0.5) / 0.5
    # Your model's forward pass
    return x
```

For post-processing, especially for segmentation tasks, you might want to convert the model's output into a segmentation mask. This can be done by applying a threshold or using argmax for multi-class segmentation:

```python
def forward(self, x):
    x = self.model(x)
    # Convert logits to probabilities and apply threshold
    x = torch.sigmoid(x) > 0.5
    return x
```

Remember to adjust these snippets to fit your specific model and task. 🛠️

Exporting your model with integrated pre and post-processing to ONNX is straightforward with `torch.onnx.export()`. Just ensure your model is in evaluation mode and specify the input dimensions correctly.

If you encounter any challenges or have more questions, feel free to reach out. Happy coding, and I'm looking forward to seeing your model in action! 🚀

**Reply by assia855** [reply]
Thank you so much @glenn-jocher  for all your code snippets I will use them for sure, also  for your confirmation that my idea for converting the model  YOLO segmentation model for ONNX export are a good idea and that encourage me to invest my time into it.

**Reply by glenn-jocher** [reply]
Hey @assia855! 🌟

I'm thrilled to hear that the code snippets and guidance have been helpful for you! It's fantastic that you're diving into converting your YOLO segmentation model for ONNX export. Your approach to integrating pre and post-processing directly into the model is spot-on and will definitely streamline your deployment process. 🚀

If you're finding certain parts of the conversion challenging, don't hesitate to share specific issues or questions. Sometimes, the devil is in the details, and I'm here to help you tackle those tricky parts. Remember, persistence is key, and it sounds like you're on the right path!

Keep up the great work, and I'm looking forward to hearing more about your progress and successes with your custom model. Happy coding! 😊

**Reply by assia855** [reply]
Hi @glenn-jocher, I'm struggling in converting  the function [crop_mask](https://github.com/ultralytics/ultralytics/blob/main/examples/YOLOv8-Segmentation-ONNXRuntime-Python/main.py#L192), my pytorch conversion is as follow: 

![Screenshot from 2024-03-15 18-39-58](https://github.com/ultralytics/ultralytics/assets/62453226/7754f81f-5a47-40be-bf36-0e0c1d577813)

The code is stuck in `return masks * ((r >= x1) * (r < x2) * (c >= y1) * (c < y2))`

I can compute the rows with x1 and x2: `(r >= x1) * (r < x2)` but I have mismatch size when computing `(c >= y1) * (c < y2)`.

Any ideas to debug this error please ? 

Thanks in advance 😊.

**Reply by glenn-jocher** [reply]
Hey there! 😊 It looks like you're on the right path with merging the `crop_mask` function directly into your PyTorch model for seamless ONNX export. It's great that the segmentation script is working well for you!

For the issue with mismatched sizes in computing `(c >= y1) * (c < y2)`, it seems like a broadcasting issue. When mixing row condition `(r >= x1) * (r < x2)` and column condition `(c >= y1) * (c < y2)` ensure both are correctly broadcasted to match the `masks` tensor dimensions.

Here's a simplified approach assuming `masks` is of shape `[batch_size, height, width]`, and `r` and `c` are indices:

```python
import torch

def crop_mask(masks, x1, y1, x2, y2):
    batch, height, width = masks.size()
    rows = torch.arange(0, height).unsqueeze(1).expand(height, width)
    cols = torch.arange(0, width).unsqueeze(0).expand(height, width)
    
    # Creating a mask for rows and columns that fall within the bounding box
    row_mask = (rows >= x1) & (rows < x2)
    col_mask = (cols >= y1) & (cols < y2)
    
    # Combining row and column masks and applying it to our input masks
    final_mask = (row_mask & col_mask).unsqueeze(0)  # Adding batch dimension
    cropped_masks = masks * final_mask.float()
    
    return cropped_masks
```

This approach ensures that your row and column masks are correctly applied across each dimension of your `masks` tensor, addressing the size mismatch issue.

Remember, you must adjust the indices (x1, y1, x2, y2) and dimensions according to your `masks` tensor layout. This is a general approach and might need fine-tuning to fit your exact use case perfectly.

Keep experimenting, and don't hesitate to ask if you run into more questions. Cheers! 😊

**Reply by assia855** [reply]
Hi @glenn-jocher  thanks for the snippet of code and the explanation. I use your code and I can compute now the `row_mask & col_mass ` but the code crash while computing `cropped_masks` with the same error: 

![Screenshot from 2024-03-20 11-11-46](https://github.com/ultralytics/ultralytics/assets/62453226/06de7c98-9216-4a5b-9296-347940427b2b)

There's a mismatch between the masks with the size: `torch.Size([8400, 160, 160])` and the final mask with the size: `torch.Size([1, 8146, 160, 160])`  I don't understand why I have 8146 in seat of 8400 for the final_mask. Any idea please? 
Thanks in advance  😊

**Reply by glenn-jocher** [reply]
Hey there! 😊 It looks like you're encountering a size mismatch issue when trying to apply your final mask to the `masks` tensor. The mismatch between `[8400, 160, 160]` and `[1, 8146, 160, 160]` typically indicates a difference in the expected batch size or the number of detected objects.

This issue could arise from incorrect assumptions about dimensions or from not matching the expected batch size in your `final_mask`. Here's a quick tip to check and align their sizes:

1. **Check Batch Dimension:** Ensure that the `final_mask` batch dimension matches the masks tensor. Since your `masks` tensor has a batch size of 8400, your `final_mask` should be expanded or repeated to match this dimension.

2. **Use Broadcasting Flexibly:** PyTorch supports broadcasting, which intelligently expands dimensions when possible. However, dimensions must align or be compatible. Make sure your `final_mask` can be broadcast to the size of the `masks` tensor.

3. **Resize or Align Tensors:** If necessary, adjust the shape of your `final_mask` to match the `masks` tensor by using operations like `.expand()` or `.repeat()`.

An example correction might look like this:

```python
# Assuming final_mask initially has shape [1, 8146, 160, 160]
# And you want it to match masks.shape = [8400, 160, 160]

# Option 1: Repeat the mask (if it's a constant mask)
final_mask = final_mask.repeat(8400 // 8146, 1, 1, 1)  # Adjust dimensions as needed

# Option 2: Ensure dimensionality matches exactly what you need
# This is a placeholder, adjust according to your logic
final_mask = final_mask.expand_as(masks)  # Only works if it's expandable

# Now apply the mask
cropped_masks = masks * final_mask
```

Remember, these are just general directions and might need to be adjusted based on your specific logic and dimensions. 

I hope this helps! If you're still stuck or have other questions, feel free to ask. Keep up the great work! 😊

**Comment by shalchianmh** [comment]
Hi,
Is there any clear and brief documentation, code or paper that tells me what pre-post processing should I perform on exported onnx object detection model to use it?
thanks.

**Reply by glenn-jocher** [reply]
@shalchianmh hello,

For pre-processing and post-processing steps when using an exported ONNX object detection model, you typically need to handle image resizing, normalization, and converting model outputs like bounding boxes and class scores into a usable format.

Here's a brief overview:

1. **Pre-processing**:
   - **Resize** your input image to the model's expected input size (e.g., 640x640 for YOLOv8).
   - **Normalize** the pixel values (commonly to a 0-1 range or using specific mean and std if required by the model).

2. **Post-processing**:
   - **Decode** the outputs to extract bounding box coordinates and class probabilities.
   - Apply **Non-Max Suppression (NMS)** to filter overlapping boxes based on their confidence scores.

For specific code examples and more detailed instructions, the export section of our documentation provides insights on handling ONNX models. While the exact steps can vary slightly based on the model specifics and your application needs, these guidelines should give you a solid starting point.

Thank you for reaching out, and I hope this helps! 😊

**Comment by teengineer** [comment]
Hi,
I export my custom model with this code block
```
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("yolov8n-seg.pt")  # load an official model
model = YOLO("..../best.pt")

# # Export the model to ONNX format
model.export(format="onnx")  # creates 'yolov8n.onnx'
```
When i test this model with this code:
```
from ultralytics import YOLO

# Load the exported ONNX model
onnx_model = YOLO(".../best.onnx")

# Define the class names (replace with your actual class names)
class_names = [0] 

# Run inference
results = onnx_model("..../00001.jpg", classes=class_names, save=True)
```
it just detect object, didnt segmentation.
Can you help me?

**Reply by glenn-jocher** [reply]
@teengineer hi there,

Thank you for reaching out! It looks like you're encountering an issue where your exported ONNX model is only detecting objects but not performing segmentation. Let's troubleshoot this together.

1. **Verify Model Type**: Ensure that the model you're exporting is indeed a segmentation model. From your code, it looks like you're loading a segmentation model (`yolov8n-seg.pt`), but let's confirm that the custom model (`best.pt`) is also a segmentation model.

2. **Check Export Process**: When exporting the model, make sure that the segmentation capabilities are preserved. The export code you provided looks correct, but let's double-check:
    ```python
    from ultralytics import YOLO

    # Load the YOLOv8 segmentation model
    model = YOLO("path/to/best.pt")

    # Export the model to ONNX format
    model.export(format="onnx")  # creates 'best.onnx'
    ```

3. **Inference Code**: Ensure that the inference code is correctly set up to handle segmentation. Here’s an example:
    ```python
    from ultralytics import YOLO

    # Load the exported ONNX model
    onnx_model = YOLO("path/to/best.onnx")

    # Run inference
    results = onnx_model("path/to/00001.jpg")

    # Check if segmentation masks are present
    if results[0].masks is not None:
        print("Segmentation masks detected!")
        results[0].plot()  # Visualize the results
    else:
        print("No segmentation masks detected.")
    ```

4. **Update Packages**: Ensure you are using the latest versions of the `ultralytics` package and ONNX runtime. You can update them using:
    ```bash
    pip install --upgrade ultralytics onnxruntime
    ```

5. **Reproducible Example**: If the issue persists, providing a minimum reproducible example would be very helpful. You can follow the guidelines here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

By following these steps, we should be able to pinpoint the issue. If you continue to face challenges, please share more details or any error messages you encounter. This will help us assist you more effectively.

**Reply by teengineer** [reply]
@glenn-jocher thank you for your responding.
I tried with 'yolov8n-seg.pt', it worked, but my exported model did not work properly, even though my model was a segmentation model. 
This is my code and the results of the codes
```
%pip install --upgrade ultralytics onnxruntime
```
```
from google.colab import drive
drive.mount('/content/drive')
```
```
from ultralytics import YOLO

# Load custom model
model = YOLO("/content/drive/path/best.pt")

# # Export the model to ONNX format
model.export(format="onnx")
```
```
from ultralytics import YOLO

# Load the exported ONNX model
onnx_model = YOLO("/content/drive/path/best.onnx")

class_names = [0]

# Run inference
results = onnx_model("/content/drive/path/00001.jpg", task='segment', classes=class_names, save=True)

# Check if segmentation masks are present
if results[0].masks is not None:
    print("Segmentation masks detected!")
    results[0].plot()  # Visualize the results
else:
    print("No segmentation masks detected.")
```
The result: 
![image](https://github.com/ultralytics/ultralytics/assets/63187528/b58ad772-1230-4906-90d1-8df88f8150d4)

```
from ultralytics import YOLO

# Load a model
model = YOLO('/content/drive/path/best.pt')  # load a custom model

results = model('/content/drive/path/00001.jpg', save=True)  # predict on an image
```
The result: 
![image](https://github.com/ultralytics/ultralytics/assets/63187528/94badbb5-2edd-4b15-88e2-5e059f51286f)

**Reply by glenn-jocher** [reply]
Hi @teengineer,

Thank you for providing the detailed code and results! It's great to see that you're making progress with the ONNX export. Let's address the issue where your custom model isn't performing segmentation as expected after export.

### Steps to Troubleshoot:

1. **Verify Model Type**: Ensure that your custom model (`best.pt`) is indeed a segmentation model. You can check this by running inference with the PyTorch model before exporting:
    ```python
    from ultralytics import YOLO

    # Load the custom model
    model = YOLO("/content/drive/path/best.pt")

    # Run inference to verify segmentation
    results = model("/content/drive/path/00001.jpg", task='segment', save=True)
    if results[0].masks is not None:
        print("Segmentation masks detected!")
    else:
        print("No segmentation masks detected.")
    ```

2. **Export Process**: Ensure the export process is correctly preserving the segmentation capabilities:
    ```python
    from ultralytics import YOLO

    # Load the custom model
    model = YOLO("/content/drive/path/best.pt")

    # Export the model to ONNX format
    model.export(format="onnx")  # creates 'best.onnx'
    ```

3. **Inference with ONNX Model**: When running inference with the ONNX model, make sure to specify the task as 'segment':
    ```python
    from ultralytics import YOLO

    # Load the exported ONNX model
    onnx_model = YOLO("/content/drive/path/best.onnx")

    # Run inference
    results = onnx_model("/content/drive/path/00001.jpg", task='segment', save=True)

    # Check if segmentation masks are present
    if results[0].masks is not None:
        print("Segmentation masks detected!")
        results[0].plot()  # Visualize the results
    else:
        print("No segmentation masks detected.")
    ```

### Additional Tips:
- **Update Packages**: Ensure you are using the latest versions of the `ultralytics` and `onnxruntime` packages:
    ```bash
    pip install --upgrade ultralytics onnxruntime
    ```

- **Reproducible Example**: If the issue persists, providing a minimum reproducible example would be very helpful. You can follow the guidelines here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

By following these steps, we should be able to pinpoint the issue. If you continue to face challenges, please share more details or any error messages you encounter. This will help us assist you more effectively.

**Comment by fukailonga** [comment]
Hi @frtendero & @glenn-jocher，I am using the following code to export the segmentation onnx model：
from ultralytics import YOLO
model = YOLO('best.pt')
success = model.export(format='onnx',opset=11,simplify=True)
And I want to run onnx through the following code:
from ultralytics import YOLO
onnx_model = YOLO("best.onnx")
results = onnx_model("D:/yolov8_lawn/ultralytics-8.2.58/ultralytics/data/images")
if results[0].masks is not  None:
    print("Segmentation masks detected！")
    results[0].plots()   # Visualize the results
else:
    print("No segmentation masks detected")
But I encountered the following problem during runtime：AttributeError: module 'onnxruntime' has no attribute 'InferenceSession'
I really hope you can help provide some guidance and suggestions,Thanks

**Reply by glenn-jocher** [reply]
Hi @fukailonga, it looks like the issue might be related to the ONNX Runtime installation. Please ensure you have the latest version of `onnxruntime` installed by running `pip install onnxruntime`. If the problem persists, try updating your Ultralytics package as well. For more detailed guidance, you can refer to our [ONNX export documentation](https://docs.ultralytics.com/integrations/onnx/). Let me know if you need further assistance!

**Comment by ishan21sh** [comment]
```
from ultralytics import YOLO

# Load the YOLOv8 model
model = YOLO("best_all.pt")

result_model = model('test_images/onnx_image_test.jpg',save=True)

# Export the model to ONNX format
model.export(format="onnx")  # creates 'yolov8n.onnx'

# Load the exported ONNX model
onnx_model = YOLO("best_all.onnx")

# Run inference
results = onnx_model("test_images/onnx_image_test.jpg',save=True")

```

when i am running my pretained model it is working perfectly fine but when i try to run the inference using the exported onnx model it throws me a bunch of errors

```
\ultralytics\engine\results.py", line 647, in verbose
    log_string += f"{n} {self.names[int(c)]}{'s' * (n > 1)}, "
                         ~~~~~~~~~~^^^^^^^^
KeyError: 23

```

**Reply by glenn-jocher** [reply]
@ishan21sh it seems like the issue might be related to the class names not being properly exported. Please ensure you are using the latest version of the Ultralytics package. If the problem persists, try manually setting the class names in the ONNX model or refer to the [ONNX export documentation](https://docs.ultralytics.com/integrations/onnx/) for further guidance.

**Comment by BaoNguyenz** [comment]
from ultralytics import YOLO

model = YOLO('/content/drive/MyDrive/Colab Notebooks/best.pt')

export_path = model.export(format='onnx',
                           half = True,
                           dynamic=True
                           )
print(f"export path save tp: {export_path}")

AssertionError: half=True not compatible with dynamic=True, i.e. use only one.

I can't run it when export onnx file dynamic batch float 16 with YOLOv8

**Reply by glenn-jocher** [reply]
@BaoNguyenz the error occurs because `half=True` and `dynamic=True` cannot be used together when exporting to ONNX. Please choose one of these options to proceed with the export.

**Comment by andriansandi** [comment]
I am exporting YOLOV8 model to ONNX. If I am using YOLO the prediction was what I'm expecting. But when use the onnxruntime seems the output results bit off. 

Here's my code

```mport onnxruntime as ort
import cv2
import numpy as np
from PIL import Image

# Load ONNX model
onnx_model_path = "runs/detect/train3/weights/best.onnx"
session = ort.InferenceSession(onnx_model_path)

# Define class names manually (update this if you have more classes in your model)
class_names = ["healthy", "unhealthy"]  # Update this based on your actual model classes

# Load and preprocess the image
image_path = "unhealthy2.jpg"
image = Image.open(image_path)
image = np.array(image)
image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)  # Convert RGB to BGR

# Resize image to match YOLO input size (640x640 for YOLOv8)
input_size = (640, 640)
image_resized = cv2.resize(image, input_size)

# Normalize the image (YOLO expects [0, 1] values)
image_resized = image_resized / 255.0  # Normalize pixel values
image_resized = image_resized.transpose(2, 0, 1)  # Convert from HWC to CHW
image_resized = np.expand_dims(image_resized, axis=0).astype(np.float32)  # Add batch dimension

# Run inference
input_name = session.get_inputs()[0].name
outputs = session.run(None, {input_name: image_resized})

# Extract and process the detections
detections = outputs[0][0]  # Remove batch dimension

print(detections)

boxes = detections[:, :4]  # Bounding boxes (x1, y1, x2, y2)
confidences = detections[:, 4]  # Confidence scores
class_ids = detections[:, 5].astype(int)  # Class indices

# Print out the class indices to see the unexpected output
print("Class IDs returned by the model:", class_ids)

# Apply threshold (e.g., 0.5) to remove low-confidence detections
confidence_threshold = 0.5
high_confidence_idx = confidences > confidence_threshold

# Filter out low-confidence predictions
boxes = boxes[high_confidence_idx]
confidences = confidences[high_confidence_idx]
class_ids = class_ids[high_confidence_idx]

# If we have any valid detections
if len(confidences) > 0:
    # Extract the best prediction (highest confidence)
    best_idx = np.argmax(confidences)

    x1, y1, x2, y2 = map(int, boxes[best_idx])
    conf = float(confidences[best_idx])
    cls = int(class_ids[best_idx])

    # Check if the class index is valid
    if cls >= len(class_names):
        print(f"Invalid class index: {cls}, it should be between 0 and {len(class_names) - 1}")
    else:
        label = class_names[cls]

        # Store results
        dataPrediction = {
            "label": label,
            "confidence": round(conf * 100, 2),  # Convert to percentage
            "coordinates": {"x1": x1, "y1": y1, "x2": x2, "y2": y2}
        }

        # Print results
        print("LABEL:", dataPrediction["label"])
        print("CONFIDENCE:", dataPrediction["confidence"])
        print("COORDINATES:", dataPrediction["coordinates"])
else:
    print("No valid detections found.")```

**Reply by glenn-jocher** [reply]
@andriansandi thanks for sharing your code! The discrepancy likely stems from differences in preprocessing or output parsing between YOLO and ONNX Runtime. Ensure your input normalization (BGR vs. RGB, 0-1 vs. 0-255) matches exactly what YOLO expects, and verify your output tensor interpretation aligns with YOLOv8's output format. For reference implementations, see our ONNX inference examples at https://docs.ultralytics.com/integrations/onnx.

**Comment by NasserAlqeifi** [comment]
why the masks are none , in the ONNX model that I export from pt custom model ? 

ultralytics.engine.results.Results object with attributes:

boxes: ultralytics.engine.results.Boxes object
keypoints: None
masks: None
names: {0: 'Bicycle lane', 1: 'Manhole', 2: 'Person', 3: 'Pothole', 4: 'Traffic Light', 5: 'Trash Can', 6: 'Vehicle', 7: 'crosswalk', 8: 'curb', 9: 'dog', 10: 'grass', 11: 'pedestrian area', 12: 'pole', 13: 'road', 14: 'sidewalk', 15: 'train'}
obb: None
orig_img: array([[[252, 253, 255],
        [254, 254, 255],
        [253, 252, 255],
        ...,
        [ 40,  34,  18],
        [ 40,  34,  20],
        [ 40,  34,  20]],

       [[254, 251, 255],
        [253, 252, 255],
        [255, 250, 253],
        ...,
        [ 37,  31,  18],
        [ 37,  31,  17],
        [ 37,  31,  17]],

       [[255, 254, 255],
        [255, 252, 255],
        [255, 251, 254],
        ...,
        [ 35,  29,  15],
        [ 35,  29,  15],
        [ 35,  29,  15]],

       ...,

       [[ 80,  81,  85],
        [ 80,  81,  85],
        [ 80,  81,  85],
        ...,
        [ 65,  65,  63],
        [ 65,  65,  63],
        [ 65,  65,  63]],

       [[ 79,  80,  84],
        [ 79,  80,  84],
        [ 79,  80,  84],
        ...,
        [ 65,  65,  63],
        [ 65,  65,  63],
        [ 65,  65,  63]],

       [[ 79,  80,  84],
        [ 79,  80,  84],
        [ 79,  80,  84],
        ...,
        [ 65,  65,  63],
        [ 65,  65,  63],
        [ 65,  65,  63]]], dtype=uint8)
orig_shape: (720, 1280)
path: 'image0.jpg'
probs: None
save_dir: 'runs\\detect\\track6'
speed: {'preprocess': 3.0138999999991256, 'inference': 16.734400000000704, 'postprocess': 2.837100000000703}

**Reply by glenn-jocher** [reply]
@NasserAlqeifi to resolve missing masks in your exported ONNX model, ensure you include the `task='segment'` argument during export for segmentation models. For example:

```python
from ultralytics import YOLO
model = YOLO('custom.pt')
model.export(format='onnx', task='segment')
```

This enables mask output in the ONNX model. For more details, see the ONNX export documentation.

**Comment by bastinj98** [comment]
Hi, 
I trained d a YOLOV8-segn model to detect and segment a single class (nc=1) and then exported in to onnx (simplify, nms, half are all False) with opset = 17. 
The output shape is 1x37x8400. 

I am building a C++ pipeline to use this model to run inference. How can I process this output to get the final segmentation mask ?
